File: skills/code-architect/references/code-examples/swift/Facade.swift
// Facade — a simplified interface over a complex subsystem.

// Subsystem components.
struct Decoder {
    func decode(_ raw: String) -> String { "decoded(\(raw))" }
}

struct Normalizer {
    func normalize(_ input: String) -> String { input.lowercased() }
}

struct Encoder {
    func encode(_ input: String) -> String { "<\(input)>" }
}

// Facade hides subsystem wiring behind one method.
struct MediaPipeline {
    private let decoder = Decoder()
    private let normalizer = Normalizer()
    private let encoder = Encoder()

    func process(_ raw: String) -> String {
        let decoded = decoder.decode(raw)
        let normalized = normalizer.normalize(decoded)
        return encoder.encode(normalized)
    }
}

func runDemo() {
    let pipeline = MediaPipeline()
    print(pipeline.process("RAW_INPUT"))
}

runDemo()

File: skills/code-architect/references/code-examples/python/facade.py
"""Facade — a simplified interface over a complex subsystem."""


class Encoder:
    def encode(self, source: str) -> str:
        return f"encoded({source})"


class Muxer:
    def mux(self, stream: str) -> str:
        return f"muxed({stream})"


class Writer:
    def write(self, container: str) -> str:
        return f"written({container})"


class MediaFacade:
    """Hides the subsystem wiring behind one convenient method."""

    def __init__(self) -> None:
        self._encoder = Encoder()
        self._muxer = Muxer()
        self._writer = Writer()

    def convert(self, source: str) -> str:
        encoded = self._encoder.encode(source)
        muxed = self._muxer.mux(encoded)
        return self._writer.write(muxed)


if __name__ == "__main__":
    print(MediaFacade().convert("clip.raw"))
